src/index.ts

```diff
@@ -87,9 +87,26 @@ export const validatorCompiler: FastifySchemaCompiler<ZodAny> =
     }
   };
 
-export const serializerCompiler: FastifySerializerCompiler<ZodAny> =
-  ({ schema }) =>
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+function hasOwnProperty<T, K extends PropertyKey>(obj: T, prop: K): obj is T & Record<K, any> {
+  return Object.prototype.hasOwnProperty.call(obj, prop);
+}
+
+function resolveSchema(maybeSchema: ZodAny | { properties: ZodAny }): Pick<ZodAny, 'safeParse'> {
+  if (hasOwnProperty(maybeSchema, 'safeParse')) {
+    return maybeSchema;
+  }
+  if (hasOwnProperty(maybeSchema, 'properties')) {
+    return maybeSchema.properties;
+  }
+  throw new Error(`Invalid schema passed: ${JSON.stringify(maybeSchema)}`);
+}
+
+export const serializerCompiler: FastifySerializerCompiler<ZodAny | { properties: ZodAny }> =
+  ({ schema: maybeSchema }) =>
   (data) => {
+    const schema: Pick<ZodAny, 'safeParse'> = resolveSchema(maybeSchema);
```

Comment:
for some reason on replies we receive ZodAny wrapped inside an additional object, and that was breaking things. Now we handle both scenarios

Reply:
IIRC response schemas are a map of statusCode => schema, thus the extra depth.

Reply:
thaaaaaat explains it